lib/ASTGen/Sources/ASTGen/Attrs.swift

```diff
@@ -0,0 +1,1132 @@
+//===--- Attrs.swift ------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2022-2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import ASTBridging
+import BasicBridging
+import SwiftDiagnostics
+@_spi(ExperimentalLanguageFeatures) @_spi(RawSyntax) import SwiftSyntax
+
+extension ASTGenVisitor {
+  struct DeclAttributesResult {
+    var attributes: BridgedDeclAttributes
+    var staticSpelling: BridgedStaticSpelling
+    var staticLoc: BridgedSourceLoc
+  }
+
+  func generateDeclAttributes(_ node: some WithAttributesSyntax & WithModifiersSyntax, allowStatic: Bool) -> DeclAttributesResult {
+    var attrs = BridgedDeclAttributes()
+    var staticSpelling: BridgedStaticSpelling = .none
+    var staticLoc: BridgedSourceLoc = nil
+
+    // Comments.
+    if let firstTok = node.firstToken(viewMode: .sourceAccurate) {
+      var pos = firstTok.position
+      for p in firstTok.leadingTrivia {
+        switch p {
+        case .docLineComment, .docBlockComment:
+          let range = self.generateCharSourceRange(start: pos, length: p.sourceLength)
+          let attr = BridgedRawDocCommentAttr.createParsed(self.ctx, range: range)
+          attrs.add(attr.asDeclAttribute)
+        default:
+          break
+        }
+        pos = pos.advanced(by: p.sourceLength.utf8Length)
+      }
+    }
+
+    func addAttribute(_ attr: BridgedDeclAttribute?) {
+      guard let attr else {
+        return
+      }
+      // FIXME: Diagnostics for duplicated attrs.
+      attrs.add(attr)
+    }
+
+    // '@' attributes.
+    for node in node.attributes {
+      switch node {
+      case .attribute(let node):
+        addAttribute(self.generateDeclAttribute(attribute: node))
```

Comment:
Any reason why this is `generateDeclAttribute` and below we have `generate(declModifier:)`. Seems inconsistent to me.

Reply:
`generate(xxx: XxxSyntax) -> BridgedXxx` is the default form (e.g. `generate(expr: ExprSyntax) -> BridgedExpr`) , so `DeclModifierSyntax` follows it. But for attributes, since we share `AttributeSyntax` for representing `swift::DeclAttribute` and `swift::TypeAttribute` we should clarify what we're generating.
```swift
func generateDeclAttribute(attribute: AttributeSyntax) -> BridgedDeclAttribute
func generateTypeAttribute(attribute: AttributeSyntax) -> BridgedTypeAttribute
```

Reply:
Ah, OK. Makes sense 👍🏽